Enable nxp usb host controllers

MarkWangChinese opened the pull request:
Create the mcux usb host controller driver (uhc_mcux.c) based on NXP SDK USB Host controller drivers (usb_host_ehci.c, usb_host_khci.c etc).
Support EHCI and KHCI controllers. Basic setup transfers (set_address and get_device_descriptor) are verified on frdm_k22f and mimxrt1060_evk boards.

danieldegrasse commented on modules/hal_nxp/usb/usb_host_mcux_drv_port.h:
```diff
@@ -0,0 +1,290 @@
+	host_callback_t deviceCallback;	/*!< Device attach/detach callback*/
+} usb_host_instance_t;
+
+usb_status_t USB_HostDetachDevice(usb_host_handle hostHandle, uint8_t hubNumber,
```
It might be worth adding a comment here highlighting that these functions are called by the USB host driver in the HAL, to make their usage and naming easier to follow

danieldegrasse commented on soc/arm/nxp_kinetis/k6x/soc.c:
```diff
@@ -111,9 +112,10 @@ static ALWAYS_INLINE void clock_init(void)
+#if CONFIG_USB_KINETIS || CONFIG_UDC_KINETIS || CONFIG_USB_UHC_NXP_KHCI
 	CLOCK_EnableUsbfs0Clock(kCLOCK_UsbSrcPll0,
 				DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency));
+	SYSMPU_Enable(SYSMPU, 0);
```
Might want a comment here highlighting that the SYSMPU must be enabled for USB to function. I would also only call this if `CONFIG_ARM_MPU` is not set.

danieldegrasse reviewed (commented):
Thanks for the work here!

MarkWangChinese commented on drivers/usb/uhc/uhc_mcux.c:
```diff
@@ -0,0 +1,718 @@
+
+void *OSA_MemoryAllocate(uint32_t memLength)
+{
+	void *p = (void *)k_heap_alloc(&mcux_uhc_alloc_pool, memLength, K_NO_WAIT);
```
Thanks @danieldegrasse for all the comments, I will update them. For this comment, actually I am using cacheable pool. All the buffers of OSA_MemoryAllocate of NXP USB controllers (EHCI, KHCI, IP3516, OHCI) can be cacheable. Do you know other modules that will use OSA_MemoryAllocate too in Zephyr?
If there are other modules that will use OSA too, I agree with you. I think we can add fsl_os_abstraction_zephyr.c and Kconfig to configure the OSA_MemoryAllocate's heap pool size and whether it is cacheable. And move all other OSA related (event, mutex) functions to fsl_os_abstraction_zephyr.c and fsl_os_abstraction.h in modules/hal_nxp/.

MarkWangChinese replied:
Hi @danieldegrasse  I change the OSA related APIs to the fsl_os_abstraction.h and fsl_os_abstraction_zephyr.c, please help to review. Thanks.
I have updated based on all your other comments.

danieldegrasse replied:
Great, thank you- added some additional review comments to the PR

danieldegrasse commented on boards/arm/frdm_k64f/frdm_k64f.dts:
```diff
@@ -202,6 +202,13 @@ zephyr_udc0: &usbotg {
 	num-bidir-endpoints = <8>;
 };
 
+/* zephyr_uhc0: &uhc1 {
```
Why is this commented out?

MarkWangChinese replied:
I see the auto-test uses the frdm_k64f board + max3421e. If I enable the zephyr_uhc0 in frdm_k64f device tree default, the auto-test case fail because there are two zephyr_uhc0.
So I enable the frdm_k22f too in this pr.

danieldegrasse replied:
Ah- I see. Maybe we need to make the auto-test case for the frdm_k64f board + max3421e into one run via an overlay file. See here for an example of a testcase like this in Zephyr: https://github.com/zephyrproject-rtos/zephyr/blob/9f7695dda0c3068ab6738a9f575a1c6dd363126d/tests/drivers/w1/w1_api/testcase.yaml#L24

MarkWangChinese replied:
I don't know where to find the USB test related config file. Could we resolve it by future PR not this one? Or I can delete the frdm_k64f support in this PR?

danieldegrasse replied:
We can resolve it in a future PR, and remove the K64F support from this PR. If you'd like to create an issue highlighting this and assign it to me, feel free to do so

MarkWangChinese replied:
The new issue is here https://github.com/zephyrproject-rtos/zephyr/issues/70013, I cannot assign it to you, I cannot change 'Assignees'.

danieldegrasse commented on boards/arm/frdm_k22f/frdm_k22f.dts:
```diff
@@ -174,6 +174,12 @@ zephyr_udc0: &usbotg {
 	num-bidir-endpoints = <8>;
 };
 
+zephyr_uhc0: &uhc1 {
```
We should not need to set the compatible or controller index here at the board level, these are SOC level properties. Same comments apply to other SOCs enabled in this PR.

Also, we do have a K22 SOC enabled in tree with a SYSMPU instance: https://github.com/zephyrproject-rtos/zephyr/blob/daf7c80c1599bcce969877b7d451e66496cf28e5/dts/arm/nxp/nxp_k22fx512.dtsi#L1. We should add the SYSMPU setup code to the K22 init if this SOC Kconfig is set (`CONFIG_SOC_MK22F12`)

danieldegrasse commented on drivers/usb/uhc/Kconfig.mcux:
```diff
@@ -0,0 +1,43 @@
+config USB_UHC_NXP_MCUX
+	bool
+	# Hidden option to simplify MCUX uhc requirement
+	default n
```
We should not need `default n` here, it should be the default if this Kconfig is not selected.

danieldegrasse commented on modules/hal_nxp/Kconfig:
```diff
@@ -5,3 +5,11 @@
 #
 
 # file is empty and kept as a place holder if/when Kconfig is needed
```
We can drop this comment

danieldegrasse commented on modules/hal_nxp/Kconfig:
```diff
@@ -5,3 +5,11 @@
 
 # file is empty and kept as a place holder if/when Kconfig is needed
+
+config FSL_OSA_HEAP
```
I think we should add a broader Kconfig here as well, something like `CONFIG_NXP_FSL_OSA` to enable the OSA support. I also would like to namespace these Kconfigs more, something like `NXP_FSL_OSA...`, since they are NXP specific

MarkWangChinese replied:
Thanks Daniel, I have updated the codes, please have a check whether it is what you think.

danieldegrasse commented on modules/hal_nxp/CMakeLists.txt:
```diff
@@ -1,12 +1,17 @@
   add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR} hal_nxp)
   add_subdirectory_ifdef(CONFIG_USB_DEVICE_DRIVER usb)
+  add_subdirectory_ifdef(CONFIG_UHC_DRIVER usb)
+  if (CONFIG_USB_UHC_NXP_MCUX)
```
Can we use a more generic Kconfig, like `CONFIG_NXP_FSL_OSA`, to control inclusion of this file?

MarkWangChinese replied:
Thanks Daniel, I have updated the codes, please have a check.

danieldegrasse commented on soc/arm/nxp_kinetis/k2x/soc.c:
```diff
@@ -102,7 +102,7 @@ static ALWAYS_INLINE void clock_init(void)
 	CLOCK_SetSimConfig(&simConfig);
 
-#if CONFIG_USB_KINETIS || CONFIG_UDC_KINETIS
+#if CONFIG_USB_KINETIS || CONFIG_UDC_KINETIS || CONFIG_USB_UHC_NXP_KHCI
```
Please add the SYSMPU block here if `CONFIG_ARM_MPU` and `CONFIG_SOC_MK22F12` are set, see my comment above

MarkWangChinese replied:
Hi @danieldegrasse I add CONFIG_USB_UHC_NXP_KHCI like original CONFIG_UDC_KINETIS. The code snippet is used to enable the usb clock. Do you mean that I need to add `&& CONFIG_SOC_MK22F12` in this `#if`?

danieldegrasse replied:
Something like this:

```
#if CONFIG_USB_KINETIS || CONFIG_UDC_KINETIS || CONFIG_USB_UHC_NXP_KHCI
	CLOCK_EnableUsbfs0Clock(kCLOCK_UsbSrcPll0,
				CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
#if !defined(CONFIG_ARM_MPU) && defined(CONFIG_SOC_MK22F12 )
      /* Make sure SYSMPU is enabled */
      SYSMPU_Enable(SYSMPU, 0);
#endif
#endif
```

MarkWangChinese replied:
Thanks Daniel, I have removed the codes of disable MPU. I use the same way as device in the include/zephyr/arch/arm/mpu/nxp_mpu.h file.

MarkWangChinese commented on dts/bindings/usb/nxp,uhc-khci.yaml:
```diff
@@ -0,0 +1,13 @@
+include: [usb-controller.yaml]
+
+properties:
+  controller-index:
```
Thanks Daniel, I have removed the controller-index and use the base address to find the controller id that MCUX usb driver needs. and I have removed the compatible of board level.

MarkWangChinese commented on modules/hal_nxp/usb/usb_host_mcux_drv_port.h:
```diff
@@ -0,0 +1,303 @@
+ * Copyright 2024 NXP
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
```
Hi @danieldegrasse @dleach02 This file's content is from the MCUX SDK USB host stack files, so I use MCUX SDK license BSD-3. But the auto-check of zephyr report errors. How do I handle it? Thanks.

danieldegrasse replied:
Could we keep the file in the MCUX HAL? Otherwise I believe we would need a specific exemption to include this file in the Zephyr tree

MarkWangChinese replied:
Move it to hal_nxp.

danieldegrasse commented on boards/nxp/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -186,6 +186,30 @@ zephyr_udc0: &usb1 {
+
+&usbphy1 {
+	status = "okay";
+	D_CAL = <12>;
```
We should use property names here that are lowercase with dashes, like `d-cal`. This aligns with the [Linux devicetree coding style](https://docs.kernel.org/devicetree/bindings/dts-coding-style.html#naming-and-valid-characters)

danieldegrasse commented on drivers/usb/uhc/uhc_mcux.c:
```diff
@@ -0,0 +1,720 @@
+		} else {
+			pipe_init.direction = USB_OUT;
+		}
+		pipe_init.numberPerUframe = 0; /* TODO: need right way to implement it. */
```
Could you add a broader comment here explaining:
- what we are working around here
- why this workaround is required

Is the issue that the UHC stack does not provide ways to set these parameters?

danieldegrasse commented on dts/bindings/usb/nxp,usbphy.yaml:
```diff
@@ -0,0 +1,30 @@
+include: base.yaml
+
+properties:
+  D_CAL:
```
Same comment I made on the devicetree source itself- can we use lowercase property names here, with dashes?

danieldegrasse commented on modules/hal_nxp/Kconfig:
```diff
@@ -4,4 +4,17 @@
+
+config NXP_FSL_OSA_HEAP
+	bool
+	default n
```
No need for `default n` here- Kconfigs aren't enabled by default

danieldegrasse commented on dts/bindings/usb/nxp,uhc-ehci.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright  2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: NXP USB host controller
```
Can we add the IP name to these descriptions, so something like `NXP EHCI USB host controller`?

danieldegrasse commented on drivers/usb/uhc/uhc_mcux.c:
```diff
@@ -0,0 +1,720 @@
+	return 0;
+}
+
+usb_status_t USB_HostAttachDevice(usb_host_handle hostHandle,
```
I believe I made this comment earlier but I wanted to suggest it again- could we add comments above each of the `USB_Hostxxx` functions, like the following:

```
/*
 * Note- this function will be called by the MCUX SDK host controller layer when...
 */
```

For example:
```
/*
 * Note- this function will be called by the MCUX SDK host controller layer when a USB device is attached
 */
```

I think this will help future developers working with the driver understand the use of these functions, since the fact that they are callbacks is not immediately obvious.

danieldegrasse reviewed (commented):
A few additional comments from my side- thanks again for all the work here!

MarkWangChinese commented:
> A few additional comments from my side- thanks again for all the work here!

Hi Daniel, appreciate your comments, I have fixed all your new comments.

danieldegrasse commented:
@mmahadevan108 could you take a look at this?

tmon-nordic commented:
Are the commits in this PR bisectable? Isn't the nxp hal commit supposed to go before the drivers are introduced?

dleach02 commented:
@jfischer-no can you provide some feedback here.

MarkWangChinese committed:
dts: nxp: add usb uhc support for frdm_k22f and rt1060

 - add usb host controller items to device tree
 - add usb phy items to device tree.
 - add yaml support for usb controller device tree items.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
modules: hal_nxp: improve the fsl_os_abstraction.h for usb host support

 - add NXP_FSL_OSA to control OSA
 - add event
 - add mutex
 - add OSA_MemoryAllocate
   define NXP_FSL_OSA_HEAP to enable the OSA heap.
   the different modules can select NXP_FSL_OSA_HEAP to enable it.
   define NXP_FSL_OSA_HEAP_CACHEABLE to enable cacheable heap.
   the different modules can select NXP_FSL_OSA_HEAP_CACHEABLE
   to enable it.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
drivers: uhc: implement nxp mcux uhc driver

 - uhc_mcux.c is based on SDK USB Host controller driver.
 - support EHCI (usb_host_ehci.c) and KHCI (usb_host_khci.c) now.
 - add related Kconfig.
 - add usb_host_config.h
 - update modules/hal_nxp and modules/hal_nxp/usb cmake file for usb host.
 - update hal_nxp revision to contain the support of usb host controllers.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
boards: nxp: add uhc suport for frdm_k22f and rt1060

 - add uhc related items to dts.
 - add clock initialization
 - add BM4 if CONFIG_USB_UHC_NXP_KHCI is enabled

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

butok commented on modules/hal_nxp/fsl_os_abstraction_zephyr.c:
```diff
@@ -0,0 +1,63 @@
+
+#ifdef CONFIG_NXP_FSL_OSA_HEAP
+/* calculate the required heap size of mcux usb host controller */
+#ifdef CONFIG_USB_UHC_NXP_MCUX
```
OSA should not have IP specific code  or configuration.
Please remove the USB code from the Common OSA, and find other solution.

MarkWangChinese replied:
Thanks Andrej, My thought is: The osa can define the heap size based on all the enabled modules that uses the osa heap.  Do you have any suggestions?

butok replied:
It should be a project configuration parameter.

MarkWangChinese replied:
I think it is better to put the similar logic to Kconfig as follow.
```
config NXP_FSL_OSA_HEAP_SIZE
	int
	default 2500 if CONFIG_USB_UHC_NXP_MCUX && CONFIG_NXP_OTHER_MODULE
	default 1500 if CONFIG_USB_UHC_NXP_MCUX
	default 1000 if CONFIG_NXP_OTHER_MODULE
	default 0
```
since there is no CONFIG_NXP_OTHER_MODULE now. the code is:
```
config NXP_FSL_OSA_HEAP_SIZE
	int
	default 1500 if CONFIG_USB_UHC_NXP_MCUX
	default 0
```

What do you think?

butok replied:
Can we use already existing Zephyr heap definitions?

MarkWangChinese replied:
Do you mean the CONFIG_HEAP_MEM_POOL_ADD_SIZE_ ${DOTCONFIG}? Different modules can increase the heap size by module's self macro. If we don't need the osa heap to be non-cachable (usb controller driver doesn't need it now) I think we can do it. In current USB stack, the system heap (k_malloc) is not enabled default, so I need to enable it if nxp usb controller is enabled. I can change the PR for review if this is your suggestion.

MarkWangChinese replied:
Thanks Andrej

Radimli commented on modules/hal_nxp/fsl_os_abstraction.h:
1) There is already fsl_os_abstraction.h in our SDK repository. This file is supposed to be independent and all implementation should be put inside fsl_os_abstraction_zephyr.h/c. 
Do you want to achieve higher speed using macros? 

2) Why don't you want to include fsl_os_abstraction_zephyr.h from our SDK repository (NXP HAL) and put implementation for Zephyr abstraction in there?

3) How are you planning to maintain two APIs (fsl_os_abstraction.h) if one of them is changed? Imagine - someone will change the API in Zephyr OR SDK and this will break our middleware, since code of middlewares is shared between both Zephyr and SDK.

MarkWangChinese replied:
Thanks Radim, Firstly The fsl_os_abstraction.h and fsl_os_abstraction_zephyr.c are already there before this PR. I add mutex, event and heap support because USB host controller drivers need them. I am totally OK to replace this osa with hal_nxp/mcux/mcux-sdk/components/osa, but zephyr is still not supported in hal_nxp/mcux/mcux-sdk/components/osa.

Radimli replied:
Alright, I just noticed, there is already Zephyr port implementation in our SDK. See https://bitbucket.sw.nxp.com/projects/MCUCORE/repos/mcu-sdk-component/browse/osa
So the thing is how to get this version working with our project inside Zephyr.

jfischer-no commented:
@MarkWangChinese Please rebase and align a bit with design of the NXP UDC drivers.

@MarkWangChinese, @dleach02 Since the EHCI specification is implemented by other vendors, would you consider adding a native implementation (providing only the fsl interface)? Perhaps others are also interested on native EHCI driver. It is not a blocker given the complexity of EHCI.

MarkWangChinese commented:
> @MarkWangChinese Please rebase and align a bit with design of the NXP UDC drivers.
> 
> @MarkWangChinese, @dleach02 Since the EHCI specification is also implemented by other vendors, would you consider adding a native implementation (providing only the fsl interface)? Perhaps others are also interested on native EHCI driver. It is not a blocker given the complexity of EHCI.

@jfischer-no , Sure, I will update this PR recently (early of Sep).
For your second question, my thought is as follow: Currently the usb host stack is not mature and is only part-implemented, if we develop a native controller from scratch, it would be too complicated to stabilize both EHCI controller driver and usb host stack. NXP EHCI controller driver is already qualified with many times of NXP SDK releases, if we could use it, we could focus on the USB Host stack development. After we get a stabilized and qualified usb host stack, we could begin a native controller driver development. So, my suggestion is using the shim now and switching to native driver in future.

MarkWangChinese commented:
nxp-zephyr/zephyr repo is never used in NXP, so this PR is closed. I will use new repo to create new PR and mention it here.
